File: FSViewToImage/Classes/FSViewToImage.m
//
//  FSViewToImage.m
//  FSViewToImage
//
//  Created by FudonFuchina on 2018/11/22.
//  Copyright © 2018年 fudongdong. All rights reserved.
//

#import "FSViewToImage.h"
#import "FSUIKit.h"
#import "FSKit.h"

@implementation FSViewToImage

+ (UIImage *)imageForUIView:(nonnull UIView *)view {
    return [FSImage imageForUIView:view];
}

+ (UIImage *)screenShot {
    UIApplication *app = [UIApplication sharedApplication];
    UIView *statusBar = [app valueForKey:@"_statusBar"];
//    NSAssert(statusBar != nil, @"获取状态栏失败");
    UIImage *image1 = [self imageForUIView:statusBar];
    UIImage *image2 = [self screenShotApple];
    
    UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, UIScreen.mainScreen.bounds.size.width, UIScreen.mainScreen.bounds.size.height)];
    imageView.image = image2;
    UIImageView *sbImageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, statusBar.frame.size.width, statusBar.frame.size.height)];
    sbImageView.image = image1;
    [imageView addSubview:sbImageView];
    return [self imageForUIView:imageView];
}

+ (UIImage *)screenShotApple {
    CGSize imageSize = [[UIScreen mainScreen] bounds].size;
    UIGraphicsBeginImageContextWithOptions(imageSize, NO, 0);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    UIWindowScene *ws = FSKit.currentWindowScene;
    for (UIWindow *window in ws.windows) {
        if (![window respondsToSelector:@selector(screen)] || [window screen] == [UIScreen mainScreen]) {
            CGContextSaveGState(context);
            CGContextTranslateCTM(context, [window center].x, [window center].y);
            CGContextConcatCTM(context, [window transform]);
            CGContextTranslateCTM(context,
                                  -[window bounds].size.width * [[window layer] anchorPoint].x,
                                  -[window bounds].size.height * [[window layer] anchorPoint].y);
            
            [[window layer] renderInContext:context];
            CGContextRestoreGState(context);
        }
    }
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

@end

File: FSViewToImage/Classes/FSViewToImage.swift
//
//  FSViewToImage.swift
//  FSViewToImage
//
//  Created by FudonFuchina on 2018/11/22.
//  Copyright © 2018年 fudongdong. All rights reserved.
//

import UIKit

public class FSViewToImage {
    
    /// 将 UIView 转换为 UIImage
    public static func image(for view: UIView) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(view.bounds.size, false, 0)
        defer { UIGraphicsEndImageContext() }
        
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        view.layer.render(in: context)
        return UIGraphicsGetImageFromCurrentImageContext()
    }
    
    /// 屏幕截图
    /// - Note: iOS 13+ 使用私有 API 获取状态栏可能会被拒审，建议使用 screenShotWithoutStatusBar()
    public static func screenShot() -> UIImage? {
        // 注意：通过 KVC 获取 _statusBar 在 iOS 13+ 已不可用
        // 这里提供一个兼容方案
        if #available(iOS 13.0, *) {
            guard let windowScene = currentWindowScene else { return nil }
            
            let screenSize = UIScreen.main.bounds.size
            UIGraphicsBeginImageContextWithOptions(screenSize, false, 0)
            defer { UIGraphicsEndImageContext() }
            
            guard let context = UIGraphicsGetCurrentContext() else { return nil }
            
            for window in windowScene.windows {
                if window.screen == UIScreen.main {
                    context.saveGState()
                    context.translateBy(x: window.center.x, y: window.center.y)
                    context.concatenate(window.transform)
                    context.translateBy(
                        x: -window.bounds.size.width * window.layer.anchorPoint.x,
                        y: -window.bounds.size.height * window.layer.anchorPoint.y
                    )
                    window.layer.render(in: context)
                    context.restoreGState()
                }
            }
            
            return UIGraphicsGetImageFromCurrentImageContext()
        }
        
        return nil
    }
    
    /// 屏幕截图（不包含状态栏的私有 API 访问）
    public static func screenShotApple() -> UIImage? {
        if #available(iOS 13.0, *) {
            guard let windowScene = currentWindowScene else { return nil }
            
            let imageSize = UIScreen.main.bounds.size
            UIGraphicsBeginImageContextWithOptions(imageSize, false, 0)
            defer { UIGraphicsEndImageContext() }
            
            guard let context = UIGraphicsGetCurrentContext() else { return nil }
            
            for window in windowScene.windows {
                if window.screen == UIScreen.main {
                    context.saveGState()
                    context.translateBy(x: window.center.x, y: window.center.y)
                    context.concatenate(window.transform)
                    context.translateBy(
                        x: -window.bounds.size.width * window.layer.anchorPoint.x,
                        y: -window.bounds.size.height * window.layer.anchorPoint.y
                    )
                    window.layer.render(in: context)
                    context.restoreGState()
                }
            }
            
            return UIGraphicsGetImageFromCurrentImageContext()
        }
        
        return nil
    }
    
    // MARK: - Private
    
    @available(iOS 13.0, *)
    private static var currentWindowScene: UIWindowScene? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
    }
}
